[BUG] - Reflect panic with too many jobs possibly

### Describe the bug

<!---A clear and concise description of what the bug is.--->

reported by @Algalish

```
panic: reflect: call of reflect.Value.Type on zero Value

goroutine 44934 [running]:
reflect.Value.Type(0x0, 0x0, 0x0, 0x0, 0x175fdd5aee8)
        /usr/local/go/src/reflect/value.go:1877 +0x166
github.com/go-co-op/gocron.callJobFuncWithParams(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x175fe0375a8, 0x40fd526000000000, 0x0, 0x0)
       /projects/src/src/github.com/go-co-op/gocron/gocron.go:53 +0xc5
github.com/go-co-op/gocron.(*Job).run(0xc0000dc540)
        /projects/src/src/github.com/go-co-op/gocron/job.go:52 +0x13b
created by github.com/go-co-op/gocron.(*Scheduler).run
        /projects/src/src/github.com/go-co-op/gocron/scheduler.go:290 +0x3f
```

### To Reproduce

```
cronScheduler = gocron.NewScheduler(time.UTC)
	for h := 0; h < 24; h++ {
		for m := 0; m < 60; m += 20 {
			timeString := fmt.Sprint(h) + ":" + fmt.Sprint(m)
			cronScheduler.Every(1).Day().At(timeString).Do(getModDetails)
			log.Println(timeString)
		}
	}
```

### Version

<!---The release version or commit SHA you're using.--->



### Expected behavior

<!---A clear and concise description of what you expected to happen.--->

### Additional context

<!---Add any other context about the problem here.--->

I modified the `for` loop a little bit to debug this:
```go
for h := 0; h < 24; h++ {
	for m := 0; m < 60; m += 20 {
		timeString := fmt.Sprint(h) + ":" + fmt.Sprint(m)
		fmt.Println("timeString: ", timeString)
		if _, err := cronScheduler.Every(1).Day().At(timeString).Do(task); err != nil {
			panic(err)
		}
		log.Println(timeString)
	}
}
```
This is the output I got:

```text
timeString:  0:0
panic: time format error

goroutine 1 [running]:
main.main()
        /tmp/test/main.go:22 +0x5f9
exit status 2
```

As you can see, there is an error while running the `Do` function. But since the error is not being handled, the `panic` happens. 

This is happening because we create a placeholder job in the `Every` function. And we only assign a "task"/"function" to that job in the `Do` function. But in this case, we are returning early in the `Do` function, leaving the `job.funcs` to be empty. Later when we call  `callJobFuncWithParams`, the `jobFunc` passed to it is `nil`. And doing a `.Type()` on this `nil` is what is causing the above panic. 

The simple fix for this would be to remove this placeholder job from the scheduler while returning any error from the `Do` function. Making a PR in sometime. If there is a better way to handle this, we can discuss it in the PR.

--- a/scheduler.go
+++ b/scheduler.go
@@ -391,11 +391,16 @@ func (s *Scheduler) stopScheduler() {
 func (s *Scheduler) Do(jobFun interface{}, params ...interface{}) (*Job, error) {
 	j := s.getCurrentJob()
 	if j.err != nil {
+		// delete the job from the scheduler as this job
+		// cannot be executed
+		s.RemoveByReference(j)
 		return nil, j.err
 	}
 
 	typ := reflect.TypeOf(jobFun)
 	if typ.Kind() != reflect.Func {
+		// delete the job for the same reason as above
+		s.RemoveByReference(j)
 		return nil, ErrNotAFunction
 	}
 

--- a/scheduler_test.go
+++ b/scheduler_test.go
@@ -846,3 +846,19 @@ func TestRunJobsWithLimit(t *testing.T) {
 	assert.Exactly(t, 1, j1Counter)
 	assert.Exactly(t, 1, j2Counter)
 }
+
+func TestErrorDuringJobCreation(t *testing.T) {
+	// error due to bad time format
+	badTime := "0:0"
+	s := NewScheduler(time.Local)
+	s.Every(1).Day().At(badTime).Do(func() {})
+	assert.Zero(t, len(s.jobs), "The job should be deleted if the time format is wrong")
+
+	// error due to the arg passed to Do() not being a function
+	s.Every(1).Second().Do(1)
+	assert.Zero(t, len(s.jobs), "The job should be deleted if the arg passed to Do() is not a function")
+
+	// positive case
+	s.Every(1).Day().Do(func() {})
+	assert.Equal(t, 1, len(s.jobs))
+}

--- a/scheduler_test.go
+++ b/scheduler_test.go
@@ -135,6 +135,13 @@ func TestAtFuture(t *testing.T) {
 	nextRun = dayJob.ScheduledTime()
 	assert.Equal(t, expectedStartTime, nextRun)
 	assert.Equal(t, false, shouldBeFalse, "Day job was not expected to run as it was in the future")
+	s.RemoveByReference(dayJob)
+
+	// error due to bad time format
+	badTime := "0:0"
+	s.Every(1).Day().At(badTime).Do(func() {})
+	assert.Zero(t, len(s.jobs), "The job should be deleted if the time format is wrong")
+
 }
 
 func schedulerForNextOrPreviousWeekdayEveryNTimes(weekday time.Weekday, next bool, n uint64, s *Scheduler) *Scheduler {
@@ -847,18 +854,30 @@ func TestRunJobsWithLimit(t *testing.T) {
 	assert.Exactly(t, 1, j2Counter)
 }
 
-func TestErrorDuringJobCreation(t *testing.T) {
-	// error due to bad time format
-	badTime := "0:0"
-	s := NewScheduler(time.Local)
-	s.Every(1).Day().At(badTime).Do(func() {})
-	assert.Zero(t, len(s.jobs), "The job should be deleted if the time format is wrong")
-
-	// error due to the arg passed to Do() not being a function
-	s.Every(1).Second().Do(1)
-	assert.Zero(t, len(s.jobs), "The job should be deleted if the arg passed to Do() is not a function")
-
-	// positive case
-	s.Every(1).Day().Do(func() {})
-	assert.Equal(t, 1, len(s.jobs))
+func TestDo(t *testing.T) {
+	var tests = []struct {
+		name     string
+		evalFunc func(*Scheduler)
+	}{
+		{
+			name: "error due to the arg passed to Do() not being a function",
+			evalFunc: func(s *Scheduler) {
+				s.Every(1).Second().Do(1)
+				assert.Zero(t, len(s.jobs), "The job should be deleted if the arg passed to Do() is not a function")
+			},
+		},
+		{
+			name: "positive case",
+			evalFunc: func(s *Scheduler) {
+				s.Every(1).Day().Do(func() {})
+				assert.Equal(t, 1, len(s.jobs))
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			s := NewScheduler(time.Local)
+			tt.evalFunc(s)
+		})
+	}
 }

